chall12/vsempere.c

```diff
@@ -0,0 +1,209 @@
+#include <stdlib.h>
+#include <string.h>
+
+char *ft_reverse_parenthesis(const char *str);
+static char *reverse(const char *str, int size);
+static int ft_reverse_parenthesis_left(char *const str, int pos, int *end_block_left);
+static int ft_reverse_parenthesis_right(char *const str, int pos, int len, int *ini_block_right);
+
+static void ft_swap_blocks(char *const str, int end_left_block, int ini_right_block)
+{
+    char *tmp;
+    int len;
+
+    len = strlen(str);
+    tmp = calloc(len + 1, sizeof(char));
+    strcat(tmp, &str[len - ini_right_block]);
+    strncat(tmp, &str[end_left_block], ini_right_block - end_left_block);
+    strncat(tmp, str, end_left_block);
+    memmove(str, tmp, len);
+    free(tmp);
+    tmp = 0;
+}
+
+static int ft_look_for_left_parenthesis(const char *str, int start)
+{
+    int open_parenthesis;
+    int i;
+
+    open_parenthesis = 0;
+    i = start;
+    while (i >= 0) {
+        if (str[i] == ')')
+            open_parenthesis++;
+        else if (str[i] == '(')
+        {
+            if (!open_parenthesis)
+                return (i);
+            else
+                open_parenthesis--;
+        }
+        
+        i--;
+    }
+    return (-1);
+}
+
+static int ft_look_for_right_parenthesis(const char *str)
+{
+    int open_parenthesis;
+    int i;
+
+    open_parenthesis = 0;
+    i = 0;
+    while (str[i]) {
+        if (str[i] == '(')
+            open_parenthesis++;
+        else if (str[i] == ')')
+        {
+            if (!open_parenthesis)
+                return (i);
+            else
+                open_parenthesis--;
+        }
+        i++;
+    }
+    return (-1);
+}
+
+static char *reverse(const char *str, int size)
+{
+    int i;
+    int block_end;
+    char *ret;
+
+    if (!(ret = calloc(strlen(str) + 1, sizeof(char))))
+        return 0;
```

Comment:
It seems easier to allocate the part you want to reverse, but it will lead to many allocation, and will utlimately be slow and greedy. You might want to simply clone the string once and reverse directly each time you encounter a parentheses. Or better you could directly do the reverse while reading and cloning the input.